File: DFS&BFS/javascript/1167.js
let input = require('fs').readFileSync('/dev/stdin').toString().split('\n');

const v = Number(input[0]);
let tree = new Map();
let node;
for (let i = 1 ; i<input.length-1 ; i++){
  input[i] = input[i].split(' ').map(Number);
  node = input[i][0];
  tree.set(node,[]);
  for (let j = 1 ; j<input[i].length-1 ; j+=2){
    tree.get(node).push([input[i][j], input[i][j+1]]);
  }
}

const dfs = (node,visited,value) =>{
  visited[node] = true;
  let longestNode = node;
  let longestCost = value;
  let searchResult;
  for (const [nextNode,cost] of tree.get(node)){
    if (visited[nextNode]) continue;
    searchResult = dfs(nextNode,visited,value+cost);
    if (searchResult[1] > longestCost){
      longestNode = searchResult[0];
      longestCost = searchResult[1];
    }
  }
  return [longestNode,longestCost];
}

let visited = new Array(v+1).fill(false);
let visitedTwo = new Array(v+1).fill(false);
let [firstNode, firstCost] = dfs(1,visited,0);
let [longestNode, longestCost] = dfs(firstNode,visitedTwo,0);
console.log(longestCost);

File: DFS&BFS/python/1167.py
import sys
input = sys.stdin.readline
v = int(input())
tree = [[] for _ in range(v+1)]
for _ in range(v):
    arr = list(map(int,input().rstrip().split()))
    for i in range(1,len(arr)-1,2):
        tree[arr[0]].append((arr[i],arr[i+1]))
def longest_node_search(node,visited,value):
    visited[node] = True
    result_value = value
    result_node = node
    for next_node,cost in tree[node]:
        if not visited[next_node]:
            output = longest_node_search(next_node,visited,value+cost)
            if output[1]>result_value:
                result_node = output[0]
                result_value = output[1]

    return [result_node,result_value]
visited = [False for _ in range(v+1)]
first_node,first_node_cost = longest_node_search(1,visited,0)
visited = [False for _ in range(v+1)]
result_node,result_node_cost = longest_node_search(first_node,visited,0)
print(result_node_cost)